tutorials/6_comp_invars.md

```diff
@@ -0,0 +1,403 @@
+
+This gives us the following
+[model variation](https://github.com/ScottWe/smartace-examples/blob/master/tutorials/post-6/instrumented/cmodel_1.c):
+
+```cpp
+/* ... Contract initialization ... */
+
+while (sol_continue()) {
+  // [START] INSTRUMENTATION (LINE 250)
+  /* Reached upon initialization, and after each iteration. */
+  /* First we construct an arbitrary network. */
+  Write_Map_1(&contract_1->user_bids, Init_sol_address_t(3),
+              Init_sol_uint256_t(nd_uint256_t("bids[3]")));
+  Write_Map_1(&contract_1->user_bids, Init_sol_address_t(4),
+              Init_sol_uint256_t(nd_uint256_t("bids[4]")));
+  Write_Map_1(&contract_1->user_bids, Init_sol_address_t(5),
+              Init_sol_uint256_t(nd_uint256_t("bids[5]")));
+  sol_require(invariant(&contract_0), "Bad arrangement.");
+  /* Next, we check the property against this arbitrary neighbourhood. */
+  sol_assert(property(&contract_0, Init_sol_address_t(0)), "Violation of Prop(0).");
+  /* ... */
+  sol_assert(property(&contract_0, Init_sol_address_t(5)), "Violation of Prop(5).");
+  // [ END ] INSTRUMENTATION
+
+  /* ... Call setup ... */
+
+  /* We allow the same local neighbourhood to run the transaction. */
+  switch (next_call) { /* ... Run transactions ... */ }
+}
+```
+
+#### The Adequacy Test
+
+We can check this model by running `cmake --build . --target verify`. This shows
+that the property dose not hold. To get some more insight, let's enable logging
+and generate a witness:
+
+  * `cmake .. -DSEA_EXELOG=true`
+  * `cmake --build . --target witness`
+
+We obtain the following concise trace.
+
+```
+[Initializing contract_0 and children]
+sender [uint8]: 3
+blocknum [uint256]: 0
+timestamp [uint256]: 0
+[Entering transaction loop]
+bids[3] [uint256]: 1
+bids[4] [uint256]: 0
+bids[5] [uint256]: 0
+assert: Violation of Prop(3).
+[sea] __VERIFIER_error was executed
+```
+
+If we go to the assertion associated with `assert: Violation of Prop(3)` we find
+that `bids[3] <= maxBid` has failed. If we walk back through the trace, we find
+the root cause of this problem, namely `bids[3] [uint256]: 1`. This says that
+`bids[3]` was assigned to a value larger than the maximum bid. Clearly, this
+assignment is not feasible.
+
+### Attempt Two: A Refined Invariant
+
+From the above analysis, we can see that the counterexample was spurious. Let's
+try to refine our compositional invariant given this trace. Perhaps we can do
+this with a linear relationship between `bids[3]` and one of the program
+variables. The obvious candidate is `bids[3] <= maxBid`. As `address(3)`
+corresponds to an arbitrary client, it is likely that our invariant generalizes
+to all clients. This gives us the new predicate shown below. The loop equates to
+checking `bids[i] <= maxBid` for each client in the neighbourhood.
+
+```cpp
+int invariant(struct Manager *c0)
+{
+  sol_uint256_t maximum = c0->user_auction.user_maxBid;
+  for (int i = 0; i < 6; ++i)
```

Comment:
Can we remove the loop by assigning to `i` a nondeterministic address?

Reply:
I will take a closer look. My tentative answer is "in general, no".

If I were to inline this manually, I think I could select a non-deterministic i for the assertions, but not for the assumptions.